### PR TITLE
Mirror sync interval specified as duration string

### PR DESCRIPTION
**Pull request by cez81**
My attempt on solution for #1140. Would love to hear your comments if I'mon the right track?

Mirror.Interval is changed from int to string so that would require a migration script?

### CODE REVIEW

**Comment by strk** [comment]
Changing a table column type would require a migration, yes.
Keeping the integer but changing semantic (for example: have them interpreted as minutes) would also require a migration.

I guess you'll want admins to be able to limit the minimum frequency of migrations (if allowed at all).

File: models/repo.go [hunk 1]
```diff
@@ -811,11 +811,15 @@ func MigrateRepository(u *User, opts MigrateRepoOptions) (*Repository, error) {
 	}
 
 	if opts.IsMirror {
+		duration, err := time.ParseDuration(setting.Mirror.DefaultInterval)
```
Comment (lunny) [inline models/repo.go]: compitable?

**Review by lunny — CHANGES_REQUESTED** [review]
And it should have a minimal duration, for example `1m` ?

File: models/repo.go (hunk 1 above)
Comment (pgaskin) [reply models/repo.go]: Also, just a note, time.ParseDuration accepts negative durations. This should be checked for.

**Comment by cez81** [comment]
I have taken your comments on board and done a few changes.
* Added a new setting mirror.MinInterval with 10 minutes as default
* MinInterval must be >1 minute or defaults to 1 minute
* DefaultInterval must be > MinInterval or defaults to 8h
* Set Interval must be >MinInterval
* Changed back to storing the interval as int in the database (store as time.Duration ie nanoseconds)
* I use duration more to save me from parsing the duration string all the time

Ps. Haven't done anything with migration yet until settled on the other bits first.

File: modules/setting/setting.go [hunk 2]
```diff
@@ -896,14 +898,20 @@ please consider changing to GITEA_CUSTOM`)
...
-	} else if err = Cfg.Section("mirror").MapTo(&Mirror); err != nil {
-		log.Fatal(4, "Failed to map Mirror settings: %v", err)
 	} else if err = Cfg.Section("api").MapTo(&API); err != nil {
 		log.Fatal(4, "Failed to map API settings: %v", err)
 	}
 
-	if Mirror.DefaultInterval <= 0 {
-		Mirror.DefaultInterval = 24
+	sec = Cfg.Section("mirror")
+	Mirror.MinInterval = sec.Key("MIN_INTERVAL").MustDuration(time.Second * 55)
```
Comment (lunny) [inline modules/setting/setting.go]: why is `time.Secdon * 55`?

Comment (cez81) [reply modules/setting/setting.go]: Oops they shouldn't be there. Removed as MustDuration will return 0 on error and that will be picked up with the following if cases.

File: modules/setting/setting.go [hunk 3]
```diff
@@ -903,8 +903,8 @@ please consider changing to GITEA_CUSTOM`)
 	}
 
 	sec = Cfg.Section("mirror")
-	Mirror.MinInterval = sec.Key("MIN_INTERVAL").MustDuration(time.Second * 55)
-	Mirror.DefaultInterval = sec.Key("DEFAULT_INTERVAL").MustDuration(time.Hour * 55)
+	Mirror.MinInterval = sec.Key("MIN_INTERVAL").MustDuration()
```
Comment (cez81) [inline modules/setting/setting.go]: Hmm somethings not right with this. Doesn't read default from built in app.ini

File: models/migrations/v27.go [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+	}
+
+	var mirrors []Mirror
+	err = x.Table("mirror").Select("*").Find(&mirrors)
+	if err != nil {
+		return fmt.Errorf("Query repositories: %v", err)
+	}
+	for _, mirror := range mirrors {
+		mirror.Interval = mirror.Interval * time.Hour
+		_, err := x.Id(mirror.ID).Cols("interval").Update(mirror)
```
Comment (lunny) [inline models/migrations/v27.go]: This may needs a transaction.

File: routers/repo/setting.go [hunk 5]
```diff
@@ -111,10 +111,13 @@ func SettingsPost(ctx *context.Context, form auth.RepoSettingForm) {
 			return
 		}
 
-		if form.Interval > 0 {
+		interval, err := time.ParseDuration(form.Interval)
+		if err != nil || interval < setting.Mirror.MinInterval {
+			ctx.Handle(500, "UpdateMirror", err)
```
Comment (lunny) [inline routers/repo/setting.go]: Show a hint is better than give a 500.

**Review by lunny — CHANGES_REQUESTED** [review]

**Comment by lunny** [comment]
Otherwise LGTM

File: models/migrations/v27.go [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+	}
+	for _, mirror := range mirrors {
+		mirror.Interval = mirror.Interval * time.Hour
+		_, err := sess.Id(mirror.ID).Cols("interval").Update(mirror)
+		if err != nil {
+			return fmt.Errorf("update mirror interval failed: %v", err)
+		}
+	}
+
+	if err := sess.Commit(); err != nil {
```
Comment (lunny) [inline models/migrations/v27.go]:
```Go
return sess.Commit()
```
is more cleaner.

**Review by lunny — APPROVED** [review]

**Comment by pgaskin** [comment]
LGTM

**Pull request merged**